src/WriterRegistrar.cpp

```diff
@@ -0,0 +1,106 @@
+  static std::map<IdNameHash, FactoryInfo> Factories;
+  return Factories;
+}
+
+IdNameHash getWriterModuleHash(std::string const &FlatbufferID,
+                               std::string const &ModuleName) {
+  return std::hash<std::string>{}(FlatbufferID + ModuleName);
+}
+
+std::map<std::string, std::string> getFactoryIdsAndNames() {
+  std::map<std::string, std::string> ReturnMap;
+  auto const Factories = getFactories();
+  for (auto const &Item : Factories) {
+    ReturnMap.insert({Item.second.Id, Item.second.Name});
+  }
+  return ReturnMap;
+}
+
+ModuleFactory const &find(std::string const &FlatbufferID) {
+  auto const &Factories = getFactories();
+  auto FoundFactory = std::find_if(Factories.begin(), Factories.end(),
+                                   [&FlatbufferID](auto const &CItem) {
+                                     return CItem.second.Id == FlatbufferID;
+                                   });
+  if (FoundFactory == Factories.end()) {
+    throw std::out_of_range("Unable to find factory with the id: " +
+                            FlatbufferID);
+  }
+  return (*FoundFactory).second.FactoryPtr;
+}
+
+ModuleFactory const &find(std::string const &FlatbufferID,
+                          std::string const &ModuleName) {
+  auto const &Factories = getFactories();
+  auto CurrentHash = getWriterModuleHash(FlatbufferID, ModuleName);
+  auto FoundItem = Factories.find(CurrentHash);
+  if (FoundItem != Factories.end()) {
+    return FoundItem->second.FactoryPtr;
+  }
+  FoundItem = std::find_if(std::cbegin(Factories), std::cend(Factories),
+                           [&ModuleName](auto const &CItem) {
+                             return CItem.second.Name == ModuleName;
+                           });
+  if (FoundItem != Factories.end()) {
+    if (not FlatbufferID.empty() and FoundItem->second.Id != FlatbufferID) {
+      throw std::runtime_error(
+          "The provided flatbuffer id (" + FlatbufferID +
+          ") does not correspont to (" + FoundItem->second.Id +
+          ") that of the found writer module (" + ModuleName + ").");
+    }
+    return FoundItem->second.FactoryPtr;
+  }
+  try {
+    return find(FlatbufferID);
+  } catch (std::out_of_range &) {
+  }
+  throw std::out_of_range("Unable to find module with name \" " + ModuleName +
+                          "\"");
+}
+
+ModuleFactory const &find(IdNameHash ModuleHash) {
+  return getFactories().at(ModuleHash).FactoryPtr;
+}
+
+void clear() { getFactories().clear(); }
+
+// cppcheck-suppress unusedFunction
+void addWriterModule(std::string const &FlatbufferID,
+                     std::string const &ModuleName, ModuleFactory Value) {
+  auto &Factories = getFactories();
+  if (FlatbufferID.size() != 4) {
+    throw std::runtime_error(
+        "The number of characters in the Flatbuffer id string must be 4.");
+  }
+  auto ModuleHash = getWriterModuleHash(FlatbufferID, ModuleName);
+  if (Factories.find(ModuleHash) != Factories.end()) {
```

Comment:
If we used emplace then can all this code be replaced like we saw yesterday?
Something like `
  Factories.emplace(ModuleHash, FactoryInfo{Value, FlatbufferID, ModuleName});` might be sufficient

Reply:
Also I think you can `std::move(Value)` here to avoid a copy